compiler/rustc_ast_passes/src/ast_validation.rs

```diff
@@ -907,6 +907,12 @@ fn validate_generic_param_order(
 
 impl<'a> Visitor<'a> for AstValidator<'a> {
     fn visit_attribute(&mut self, attr: &Attribute) {
+        if !self.in_trait_impl
+            && let Some(ident) = attr.ident()
+            && ident.name == sym::do_not_recommend
+        {
+            self.session.emit_err(IncorrectDoNotRecommendLocation { span: attr.span });
```

Comment:
Should this be implemented in [`check_attr.rs`](https://github.com/rust-lang/rust/blob/master/compiler/rustc_passes/src/check_attr.rs)? That's where all the "this attribute should only be specified on X" checking is done.

Reply:
AFAICT, `rustc_parse` is not aware of the `AstValidator` struct and so I decided to keep things unchanged. Regardless, it is possible to declare `validate_attr::check_attr(&self, attr);`, modify everything accordingly and move this snippet to `check_attr` if desired.

Reply:
Sorry, I'm not quite sure I understand what you're describing - why can't this be in implemented in `check_attr.rs` like all the other checks there?

i.e. adding a match arm..
```rust
sym::do_not_recommend => self.check_do_not_recommend(hir_id, attr, span, target),
```

..and a function..

```rust
/// Checks if `#[do_not_recommend]` is applied to a impl.
    fn check_do_not_recommend(&self, attr: &Attribute, span: Span, target: Target) -> bool {
        match target {
            Target::Impl => true,
            _ => {
                self.tcx
                    .sess
                    .emit_err(errors::DoNotRecommend { attr_span: attr.span, defn_span: span });
                false
            }
        }
    }
```

Reply:
> Should this be implemented in [`check_attr.rs`](https://github.com/rust-lang/rust/blob/master/compiler/rustc_passes/src/check_attr.rs)? That's where all the "this attribute should only be specified on X" checking is done.

Sorry, got confused with the AST counterpart.

> Sorry, I'm not quite sure I understand what you're describing - why can't this be in implemented in `check_attr.rs` like all the other checks there?
> 
> i.e. adding a match arm..
> 
> ```rust
> sym::do_not_recommend => self.check_do_not_recommend(hir_id, attr, span, target),
> ```
> 
> ..and a function..
> 
> ```rust
> /// Checks if `#[do_not_recommend]` is applied to a impl.
>     fn check_do_not_recommend(&self, attr: &Attribute, span: Span, target: Target) -> bool {
>         match target {
>             Target::Impl => true,
>             _ => {
>                 self.tcx
>                     .sess
>                     .emit_err(errors::DoNotRecommend { attr_span: attr.span, defn_span: span });
>                 false
>             }
>         }
>     }
> ```

Done as requested.